server/api/orders.js

```diff
@@ -39,4 +38,161 @@ router.put('/:pid/:pqty', async (req, res, next) => {
+          orderProduct.update({
+            //depending on implementation, either:
+            //FOR FORM:
+            quantity: Number(pqty)
+            //OR other style:
+            // quantity: Number(pqty) + orderProduct.quantity
+          });
+        }
+      });
+
+      //GET ALL ORDERPRODUCTS
+      const allOrderProducts = await OrderProduct.findAll({
+        where: {
+          orderId: usersCart.id
+        }
+      });
+
+      //ACCUMULATE GRAND TOTAL
+      let grandTotal = 0;
+      allOrderProducts.forEach(orderProduct => {
+        grandTotal += orderProduct.price * orderProduct.quantity;
+      });
+
+      //UPDATE ORDER GRANDTOTAL
+      await usersCart.update({
+        grandTotal: grandTotal
+      });
+      res.sendStatus(200);
+    } else {
+      //OTHERWISE WE"RE OUT OF THE PRODUCT
+      throw new Error('Not enough product available.');
+    }
+  } catch (error) {
+    next(error);
+  }
+});
+
+//CHECKOUT
+router.get('/checkout/:userId', async (req, res, next) => {
+  try {
+    //req.user.id equivalent should be found on the frontend and then
+    //put in on the frontend route.
+
+    //CHECK IF IN STOCK
+    const order = await Order.findOne({
+      where: {
+        userId: req.params.userId,
+        isPurchased: false
+      }
+    });
+    const allProducts = await Product.findAll({});
+    const allRespectiveOrderProducts = await OrderProduct.findAll({
+      where: {
+        orderId: order.id
+      }
+    });
+
+    let allProductsWantedAvailable = true; //a flag
+    const namesAndInventory = {}; //holds names and inventory
+    allProducts.forEach(product => {
+      namesAndInventory[product.name] = product.inventory; //gets entire arsenal
+      allRespectiveOrderProducts.forEach(orderProduct => {
+        if (product.id === orderProduct.productId) {
+          if (orderProduct.quantity > product.inventory) {
+            allProductsWantedAvailable = false;
+          }
+        }
+      });
+    });
+    //IF ALL PRODUCTS WANTED ARE AVAILABLE
+    if (allProductsWantedAvailable) {
+      //CHANGE ISPURCHASED STATUS
+      await order.update({
+        isPurchased: true
+      });
+
+      //Need a new order now with default properties
+      await Order.create({userId: req.params.userId});
+
+      //NEED TO CHANGE INVENTORY OF PRODUCTS
+      allProducts.forEach(product => {
+        allRespectiveOrderProducts.forEach(orderProduct => {
+          if (product.id === orderProduct.productId) {
+            namesAndInventory[product.name] =
+              namesAndInventory[product.name] - orderProduct.quantity;
+          }
+        });
+      });
+
+      allProducts.forEach(async product => {
+        await product.update({
+          inventory: namesAndInventory[product.name]
+        });
+      });
+
+      //redirect below to a checkout confirmation page
+      res.redirect('/products');
+    } else {
+      //BLOCK ORDER AND AT THE LEAST MAYBE SEND TO OUT OF ORDER PAGE
```

Comment:
If order failed for some reason, we should stay on checkout page and serve an error.

Reply:
I have a nice solution for this. Might be too much info given but we can append as needed later.